ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs

```diff
@@ -172,98 +191,141 @@ tickHeaderState :: ConsensusProtocol (BlockProtocol blk)
+                 (ChainDepState (BlockProtocol blk'))
+     , TipInfo blk ~ TipInfo blk'
+     )
+  => HeaderStateHistory blk -> HeaderStateHistory blk'
+castHeaderStateHistory HeaderStateHistory {..} = HeaderStateHistory {
+      headerStateHistorySnapshots = castSeq castHeaderState headerStateHistorySnapshots
+    , headerStateHistoryAnchor    =         castHeaderState headerStateHistoryAnchor
     }
   where
     -- This is unfortunate. We're doing busy-work on a strict-sequence,
     -- mapping a function that actually doesn't change anything :/
     castSeq :: (a -> b) -> StrictSeq a -> StrictSeq b
-    castSeq f = Seq.fromList . map f . toList
+    castSeq f = Seq.toStrict . fmap f . Seq.getSeq
 
--- | Rewind the header state
+-- | Rewind the header state history
 --
--- This involves 'rewindChainState', and so inherits its PRECONDITION that the
--- target point must have been previously applied.
+-- PRECONDITION: the point to rewind to must correspond to a header (or
+-- 'GenesisPoint') that was previously applied to the header state history.
 --
-rewindHeaderState :: forall blk.
-                     ( BlockSupportsProtocol blk
-                     , Serialise (HeaderHash blk)
-                     , HasAnnTip blk
-                     )
-                  => TopLevelConfig blk
-                  -> Point blk
-                  -> HeaderState blk -> Maybe (HeaderState blk)
-rewindHeaderState cfg p HeaderState{..} = do
-    chainDepState' <- rewindChainDepState
-                        (Proxy @(BlockProtocol blk))
-                        (configSecurityParam cfg)
-                        p
-                        headerStateConsensus
-    return $ HeaderState {
-        headerStateConsensus = chainDepState'
-      , headerStateTips      = Seq.dropWhileR rolledBack headerStateTips
-      , headerStateAnchor    = headerStateAnchor
-      }
+-- Rewinding the header state history is intended to be used when switching to a
+-- fork, longer or equally long to the chain to which the current header state
+-- corresponds. So each rewinding should be followed by rolling forward (using
+-- 'headerStateHistoryPush') at least as many blocks that we have rewound.
+--
+-- Note that repeatedly rewinding a header state history does not make it
+-- possible to rewind it all the way to genesis (this would mean that the whole
+-- historical header state is accumulated or derivable from the current header
+-- state history). For example, rewinding a header state by @i@ blocks and then
+-- rewinding that header state again by @j@ where @i + j > k@ is not possible
+-- and will yield 'Nothing'.
+rewindHeaderStateHistory ::
+     forall blk. (BlockSupportsProtocol blk, HasAnnTip blk)
+  => Point blk
+  -> HeaderStateHistory blk -> Maybe (HeaderStateHistory blk)
+rewindHeaderStateHistory p HeaderStateHistory {..} =
+    case Seq.dropWhileR rolledBack headerStateHistorySnapshots of
+        Empty
+          | pointToWithOriginRealPoint p /=
```

Comment:
Simplify guard to `rolledBack headerStateHistoryAnchor`?

Reply:
Nice! :pray: